crates/red_knot_test/src/parser.rs

```diff
@@ -0,0 +1,467 @@
+use once_cell::sync::Lazy;
+use regex::{Captures, Regex};
+use ruff_index::{newtype_index, IndexVec};
+use rustc_hash::{FxHashMap, FxHashSet};
+
+pub(crate) fn parse<'s>(path: &'s str, source: &'s str) -> anyhow::Result<MarkdownTestSuite<'s>> {
+    let parser = Parser::new(path, source);
+    parser.parse()
+}
+
+#[newtype_index]
+struct SectionId;
+
+#[newtype_index]
+struct FileId;
+
+/// A parsed markdown file containing tests.
+///
+/// Borrows from the source string and filepath it was created from.
+#[derive(Debug)]
+pub(crate) struct MarkdownTestSuite<'s> {
+    /// Header sections.
+    sections: IndexVec<SectionId, Section<'s>>,
+
+    /// Test files embedded within the Markdown file.
+    files: IndexVec<FileId, File<'s>>,
+}
+
+impl<'s> MarkdownTestSuite<'s> {
+    pub(crate) fn tests<'m>(&'m self) -> MarkdownTestIterator<'m, 's> {
+        MarkdownTestIterator {
+            suite: self,
+            current_file_index: 0,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MarkdownTest<'m, 's> {
+    suite: &'m MarkdownTestSuite<'s>,
+    section: &'m Section<'s>,
+    files: &'m [File<'s>],
+}
+
+impl<'m, 's> MarkdownTest<'m, 's> {
+    pub(crate) fn name(&self) -> String {
+        let mut name = self.section.title.to_string();
+        let mut parent_id = self.section.parent_id;
+        while let Some(parent) = parent_id.map(|section_id| &self.suite.sections[section_id]) {
+            parent_id = parent.parent_id;
+            name.insert_str(0, " - ");
+            name.insert_str(0, parent.title);
+        }
+        name
+    }
+
+    pub(crate) fn files(&self) -> impl Iterator<Item = &'m File<'s>> {
+        self.files.iter()
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct MarkdownTestIterator<'m, 's> {
+    suite: &'m MarkdownTestSuite<'s>,
+    current_file_index: usize,
+}
+
+impl<'m, 's> Iterator for MarkdownTestIterator<'m, 's> {
+    type Item = MarkdownTest<'m, 's>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut current_file_index = self.current_file_index;
+        let mut file = self.suite.files.get(current_file_index.into());
+        let section_id = file?.section;
+        while file.is_some_and(|file| file.section == section_id) {
+            current_file_index += 1;
+            file = self.suite.files.get(current_file_index.into());
+        }
+        let files = &self.suite.files
+            [FileId::from_usize(self.current_file_index)..FileId::from_usize(current_file_index)];
+        self.current_file_index = current_file_index;
+        Some(MarkdownTest {
+            suite: self.suite,
+            section: &self.suite.sections[section_id],
+            files,
+        })
+    }
+}
+
+#[derive(Debug)]
+struct Section<'s> {
+    title: &'s str,
+    level: usize,
```

Comment:
This comes from a `usize` (the number of `#` in the header line); if I use a `u8` or `u32` here it just means I'll have to convert from `usize` (and handle the practically-impossible scenario where it doesn't fit.) Does that still seem worth it to you to avoid `usize` here?

Reply:
Overall yes, because `usize` is intended for sizes that depend on the CPU architecture and handling the error case seems easy enough when using anyhow (`try_from(...).context(...)`)